# IR 9. 매트릭스(선언 방법, 인덱싱), apply 함수, %>% 연산자, 시간비교
> "작성완료"

- toc:true
- branch: master
- badges: true
- comments: true
- [R]

---
# Introduce R 
- lecture: Introduce R_7nd week of lectures
- lecture date: 2021-10-21
- lecturer: Guebin choi
- study date: 2022-02-15
- author: Kione kim
---

In [212]:
#hide
options(jupyter.plot_scale=3)
options(jupyter.display_mimetypes=c("text/html"))
options(jupyter.rich_display=FALSE)
options(repr.plot.width=6,repr.plot.height=6,repr.plot.res=300)
options(max.print=1000)

In [67]:
library(tidyverse)

## apply를 이용한 풀이   

``` 
입력
3 4 5 1
7 7 7 2
8 1 8 2
3 3 4 5

출력은 ? 
```

**step 1: apply함수를 사용하여 각 row에 min을 구한다.**

In [9]:
a = rbind(c(3,4,5,1), c(7,7,7,2), c(8,1,8,2), c(3,3,4,5))
a

     [,1] [,2] [,3] [,4]
[1,] 3    4    5    1   
[2,] 7    7    7    2   
[3,] 8    1    8    2   
[4,] 3    3    4    5   

In [69]:
apply(a,1,min) # a= 입력 메트릭스, 1= 입력메트릭스의 각 row에 어떠한 함수를 적용

[1] 1 2 1 3

In [71]:
a %>% apply(1,min) # 위의 코드와 같음.

[1] 1 2 1 3

**step2: step1의 결과에 max를 취한다.**

In [72]:
max(apply(a,1,min))

[1] 3

In [74]:
a %>% apply(1,min) %>% max # 역시 위와 같은 코드

[1] 3

## 카드뽑기 게임에서 배울만한 프로그래밍 문법들

### 매트릭스

#### 매트릭스를 선언하는 방법

`-` 방법 1: rbind(row vector)

In [75]:
rbind(c(4,3,2,1),c(1,2,3,4))

     [,1] [,2] [,3] [,4]
[1,] 4    3    2    1   
[2,] 1    2    3    4   

`-` 방법 2: cbind(column vector)

In [82]:
cbind(c(4,3,2,1),c(1,2,3,4))

     [,1] [,2]
[1,] 4    1   
[2,] 3    2   
[3,] 2    3   
[4,] 1    4   

`참고:` 만들어진 매트릭스의 차원(dimension)을 확인하는 방법 -> dim

In [83]:
a= rbind(c(4,3,2,1),c(1,2,3,4))
a

     [,1] [,2] [,3] [,4]
[1,] 4    3    2    1   
[2,] 1    2    3    4   

In [84]:
dim(a)

[1] 2 4

In [85]:
a= cbind(c(4,3,2,1),c(1,2,3,4))
a

     [,1] [,2]
[1,] 4    1   
[2,] 3    2   
[3,] 2    3   
[4,] 1    4   

In [86]:
dim(a)

[1] 4 2

`-` 방법 3: 벡터를 선언하고 dim()를 이용하여 차원을 강제셋팅

In [88]:
a = 1:8
a

[1] 1 2 3 4 5 6 7 8

In [89]:
dim(a)

NULL

In [90]:
dim(a) <- c(2,4) # 차원을 강제 셋팅

In [92]:
dim(a)

[1] 2 4

In [93]:
a

     [,1] [,2] [,3] [,4]
[1,] 1    3    5    7   
[2,] 2    4    6    8   

In [99]:
dim(a) <- c(4,2) # 차원을 강제 셋팅
dim(a)

[1] 4 2

In [100]:
a

     [,1] [,2]
[1,] 1    5   
[2,] 2    6   
[3,] 3    7   
[4,] 4    8   

In [101]:
dim(a) <- c(3,3) # 차원이 맞지 않기에 오류

ERROR: Error in dim(a) <- c(3, 3): dims [product 9]가 객체 [8]의 길이와 일치하지 않습니다


#### 매트릭스의 원소를 뽑는 방법

In [102]:
a= rbind(c(5,1,3,4), c(4,3,5,6), c(0,1,5,8))
a

     [,1] [,2] [,3] [,4]
[1,] 5    1    3    4   
[2,] 4    3    5    6   
[3,] 0    1    5    8   

`-` 1열 뽑기

In [103]:
a[,1]

[1] 5 4 0

`-` 1열, 4열 뽑기

In [104]:
a[,c(1,4)]

     [,1] [,2]
[1,] 5    4   
[2,] 4    6   
[3,] 0    8   

`-` 2행 뽑기

In [105]:
a[2,]

[1] 4 3 5 6

`-` 1행, 3행 뽑기

In [106]:
a[c(1,3),]

     [,1] [,2] [,3] [,4]
[1,] 5    1    3    4   
[2,] 0    1    5    8   

`-` 2행 4열 뽑기

In [108]:
a[2,4]

[1] 6

`-` 1행, 3행과 4열 뽑기

In [110]:
a[c(1,3),4]

[1] 4 8

#### 매트릭스를 쓰는 이유:

- 바둑판 구조의 자료형을 표현하기 위해서 = 자료를 바둑판 구조로 저장하기 위해서

`-` 자료를 바둑판(2d-array), 나아가 큐브형태(3d-array)로 저장하면 어떠한 이점이 있는가?    그렇다면 자료를 벡터(1d-array)로 저장하면 어떠한 이점이 있는가?

- 프로그래밍 구조상의 이점
- 계산상의 이점

### apply함수

``` R
사용법: apply(매트릭스, 1, 특정함수)
-> 어떠한 특정함수가 각각의 row에 적용되어서 array로 저장된다.
- 여기서 특정함수는 "벡터 입력 - 스칼라 출력"이라 생각하자. 즉 벡터가 입력되면 스칼라가 출력되는 함수만 apply가 가능하다고 생각하자.(사실 그렇지는 않음)
````

`-` 예시

In [111]:
sum(c(3,3,2,2))

[1] 10

- (3,3,2,2)라는 벡터 입력 -> 10이라는 스칼라 출력

In [112]:
mean(c(3,3,2,2))

[1] 2.5

- (3,3,2,2)라는 벡터 입력 -> 2.5라는 스칼라 출력

In [113]:
sqrt(c(3,3,2,2))

[1] 1.732051 1.732051 1.414214 1.414214

- (3,3,2,2)라는 벡터 입력 -> (1.732051, 1.732051, 1.414214, 1.414214)라는 벡터 출력
- 조건에 해당하지 않음

`apply에 적용`

In [114]:
a

     [,1] [,2] [,3] [,4]
[1,] 5    1    3    4   
[2,] 4    3    5    6   
[3,] 0    1    5    8   

In [115]:
apply(a,1,sum)

[1] 13 18 14

In [116]:
apply(a,1,mean)

[1] 3.25 4.50 3.50

In [117]:
apply(a,1,min)

[1] 1 3 0

In [118]:
apply(a,1,max)

[1] 5 6 8

`-` 사용자가 정의한 함수에도 적용가능하다.

In [119]:
f = function(a) (min(a)-1)^2+7

In [120]:
apply(a,1,f)

[1]  7 11  8

`-` apply(매트릭스, 2, 특정함수) -> 1을 2로 바꾸면 특정함수가 row가 아닌 각각의 column에 적용되어 결과가 array로 저장된다.

In [121]:
a

     [,1] [,2] [,3] [,4]
[1,] 5    1    3    4   
[2,] 4    3    5    6   
[3,] 0    1    5    8   

In [122]:
apply(a,2,sum)

[1]  9  5 13 18

In [125]:
apply(a,2,mean)

[1] 3.000000 1.666667 4.333333 6.000000

In [123]:
apply(a,2,max)

[1] 5 3 5 8

In [126]:
apply(a,2,f)

[1]  8  7 11 16

### **%>% 연산자($\star\star\star\star\star$)**
: 매우 중요함 !

In [127]:
sum(1:10)

[1] 55

In [128]:
1:10 %>% sum()

[1] 55

In [129]:
1:10 %>% sum

[1] 55

- 위 세 개는 동일한 코드이다. 즉 ()는 생략 가능하다

In [130]:
myfun = function(x,y) x**2+y

In [134]:
myfun(3,5)

[1] 14

In [135]:
3 %>% myfun(5)

[1] 14

- 위 두 코드는 동일한 코드임

`-` 입력이 세 개인 함수

In [136]:
myfun2 = function(x,y,z) x**2+y*z

In [137]:
myfun2(3,3,2)

[1] 15

In [139]:
3 %>% myfun2(3,2)

[1] 15

- 마찬가지로, 위 두 코드는 동일함

**`-` %>%를 사용하는 이유: 합성함수 처리에 매우 유용하기 떄문**

In [145]:
3:6 %>% log %>% sqrt %>% sum %>% myfun2(1,1) %>% sqrt

[1] 4.935136

- 처리하고 싶은대로 즉각적으로 적용 가능

In [150]:
sqrt(myfun2(sum(sqrt(log(3:6))), 1,1))

[1] 4.935136

- 복잡해질 수록 처리하기 어려워지고 수정이 어려워짐

`-` 다음을 같은 방식으로 처리해보자

In [151]:
3:6 %>% log %>% myfun(3) %>% sqrt %>% sum %>% myfun2(1,1) %>% sqrt

[1] 9.180673

In [156]:
sqrt(myfun2(sum(sqrt(myfun(log(3:6),3))),1,1))

[1] 9.180673

- 처음부터 코드를 작성하면 그래도 덜 헷갈리지만, 위에 작성한 코드를 수정하려고 하면 매우 헷갈릴 수 있다.

`-` 아래의 개념차이
- $g(h(f(x)))$
- $x \to f \to h \to g$

### 시간비교

`-` 시간비교하는 방법   
: Sys.time() 사용

In [164]:
t1= Sys.time()

In [165]:
t2= Sys.time()

In [166]:
t2-t1

Time difference of 10.89629 secs

`-` 큰 매트릭스

In [172]:
sample(1:9,size=3) # 1:9까지의 벡터를 생성하고 그 중 중복을 허용하지 않고 랜덤으로 3개를 출력

[1] 4 5 2

In [174]:
sample(1:9,size=3,replace=TRUE) # 1:9까지의 벡터를 생성하고 중복을 허용하고 랜덤으로 3개를 출력

[1] 9 1 9

In [175]:
sample(1:3,size=4)

ERROR: Error in sample.int(length(x), size, replace, prob): 'replace = FALSE' 일때는 모집단보다 큰 샘플을 가질 수 없습니다


In [188]:
sample(1:3,size=4,replace=TRUE)

[1] 3 1 3 2

In [191]:
a=sample(1:9,size = 100*2000000,replace = TRUE)

In [194]:
dim(a)

NULL

In [195]:
dim(a) <- c(2000000,100)
a

      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]
 [1,] 5    7    7    9    6    9    4    9    3    7     <U+22EF> 8     2    
 [2,] 5    7    9    2    8    1    8    9    7    8     <U+22EF> 9     5    
 [3,] 8    6    2    6    9    4    5    4    8    2     <U+22EF> 7     4    
 [4,] 6    9    6    7    5    9    6    9    1    7     <U+22EF> 4     3    
 [5,] 7    4    6    3    3    8    7    2    7    1     <U+22EF> 8     6    
 [6,] 3    4    8    3    8    9    4    4    2    7     <U+22EF> 2     9    
 [7,] 3    5    7    1    2    4    5    6    4    7     <U+22EF> 7     6    
 [8,] 6    2    6    1    8    7    7    6    7    3     <U+22EF> 3     3    
 [9,] 6    5    2    6    5    4    2    1    5    1     <U+22EF> 9     9    
[10,] 7    5    6    5    1    2    9    5    8    7     <U+22EF> 7     1    
[11,] 9    3    7    5    4    9    8    8    2    9     <U+22EF> 3     9    
[12,] 7    4    9    3    2    3    4    1    7    7     <U+22EF> 4

In [199]:
t1= Sys.time()
a %>% apply(1,min) %>% max
t2= Sys.time()
t2-t1

[1] 2

Time difference of 15.50931 secs

In [210]:
t1= Sys.time()
vec <- c()
for (i in 1:2000000) vec[i] <- min(a[i,])
max(vec)
t2= Sys.time()
t2-t1

[1] 2

Time difference of 8.893018 secs

- 위에 1:2000000은 행의 수를 알고 있어야 하기 때문에 확장성이 떨어진다고 할 수 있다. 아래와 같은 코드는 행의 수를 몰라도 되기 때문에 확장성이 있다.

In [202]:
dim(a)

[1] 2000000     100

In [203]:
dim(a)[1]

[1] 2000000

In [211]:
t1=Sys.time()
vex <- c()
for (i in 1:(dim(a)[1])) vec[i] <- min(a[i,])
max(vec)
t2=Sys.time()
t2-t1

[1] 2

Time difference of 7.932614 secs

- for문이 좀 더 빠르지만, 코드는 apply가 더 간결하다.